tests/regression/lp-1862637/task.yaml

```diff
@@ -2,16 +2,56 @@ summary: /var/lib/dhcp is shared with host
 prepare: |
     snap pack test-snapd-app
     snap install --dangerous ./test-snapd-app_1_all.snap
-    if [ -d /var/lib/dhcp ]; then
-        touch /var/lib/dhcp/.canary
-    fi
 restore: |
-    if [ -d /var/lib/dhcp ]; then
-        rm -f /var/lib/dhcp/.canary
-    fi
+    rm -f /var/lib/dhcp/.canary
     snap remove test-snapd-app
     rm -f test-snapd-app_1_all.snap
 execute: |
-    if [ -d /var/lib/dhcp ]; then
+    # The snap always sees /var/lib/dhcp. 
+    test-snapd-app.sh -c 'test -d /var/lib/dhcp'
+    if [[ "$SPREAD_SYSTEM" != ubuntu-core-16-* ]]; then
+        if [ -d /var/lib/dhcp ]; then
+            # Variant 1) The host is not a core16 system and has /var/lib/dhcp
+            # Directory is bind mounted and writable when the interface is
+            # connected.
+            touch /var/lib/dhcp/.canary
+            test-snapd-app.sh -c 'test ! -f /var/lib/dhcp/.canary'
+
+            snap connect test-snapd-app:network-control
+            test-snapd-app.sh -c 'test -f /var/lib/dhcp/.canary'
+            test-snapd-app.sh -c 'echo hi > /var/lib/dhcp/.canary'
+            test "$(cat /var/lib/dhcp/.canary)" = hi
+
+            snap disconnect test-snapd-app:network-control
+            test-snapd-app.sh -c 'test ! -f /var/lib/dhcp/.canary'
+            not test-snapd-app.sh -c 'echo bye > /var/lib/dhcp/.canary'
+            test "$(cat /var/lib/dhcp/.canary)" = hi
+        else
+            # Variant 2) The host is not a core16 system but lacks /var/lib/dhcp
+            # Directory is not created on the host and is not bind mounted.
+            # Snaps always see read only, empty directory.
+            test-snapd-app.sh -c 'test ! -f /var/lib/dhcp/.canary'
+
+            snap connect test-snapd-app:network-control
+            not test-snapd-app.sh -c 'echo hi > /var/lib/dhcp/.canary'
+            test-snapd-app.sh -c 'test ! -f /var/lib/dhcp/.canary'
+
+            snap disconnect test-snapd-app:network-control
+            test-snapd-app.sh -c 'test ! -f /var/lib/dhcp/.canary'
```

Comment:
For symmetry, also do the 'hi' test here?

Reply:
Sure. I'll echo "bye" like the variant above does and make sure both are no-op.